src/WeakMap.php

```diff
@@ -191,5 +202,41 @@ private function assertValidKey($key) : void
                 throw new TypeError('WeakMap key must be an object');
             }
         }
+
+        private function setupHousekeepingOnGcRun() : void
+        {
+            $this->destructedEarly = new \stdClass();
+
+            $weakThis = \WeakReference::create($this); // allow self to be released without GC
+            $weakThisDestructedEarly = \WeakReference::create($this->destructedEarly);
+
+            $gcRuns = 0;
+            $setupDestructorFx = static function () use ($weakThis, $weakThisDestructedEarly, &$gcRuns, &$setupDestructorFx): void {
+                $destructorFx = static function () use ($weakThis, $weakThisDestructedEarly, &$gcRuns, &$setupDestructorFx): void {
+                    $gcRunsPrev = $gcRuns;
+                    $gcRuns = gc_status()['runs'];
+                    if ($gcRunsPrev !== $gcRuns) { // prevent recursion on shutdown
+                        $setupDestructorFx();
+                    }
+
+                    if ($weakThisDestructedEarly->get() === null) {
+                        return;
+                    }
```

Comment:
Removing this condition and all code related to `$this->destructedEarly` still passes all tests on PHP 7.4. Is there a test missing to reproduce here what you reported in https://github.com/php/php-src/issues/13612?

Reply:
I have reworked this part to workaround https://github.com/php/php-src/issues/13670 too.

Now the impl. is faster and workaround for https://github.com/php/php-src/issues/13612 is not needed :).